### DISCUSSION TITLE
Externals with dependencies

### DISCUSSION THREAD

**Discussion by scheibelp** [spack/spack#21636] [Ideas]
## Summary

External entries in `packages.yaml` currently do not allow specifying dependencies. This may lead to inconsistencies. This discussion proposes how to update `packages.yaml` to record this information. This will also require some edits to the concretizer, but currently the focus is on how to update the configuration schema to record the necessary information.

## The Problem

Right now, if you have a spec DAG like

```
W->Y->Z
```

And your `packages.yaml` file includes an external entry for `Y`, and the concretization chooses this instance of `Y`, then constructed spec DAG would in many cases look like

```
W->Y
```

That is, Spack will truncate all dependencies of chosen externals. Often times this is not a problem, but it could result in consistency issues, for example:

```
  W
 / \
X.  Y
 \./
  Z
```

If the external `Y` was built with a specific instance of `Z`, and Spack builds `X` with a separate instance of `Z`, there may be linking issues.

Generally, this can be avoided if

* (EDIT: clarity) For each package marked as external, all dependencies of that package are also recorded as externals
* Each external package is marked as `buildable: false`

(in that case, packages depending on Z will always use the Z that Y used)

Even with those constraints, if you have multiple instances of Y installed, e.g. `Y+foo` and `Y~foo`, and each has their own instance of Z, then setting `buildable: False` would not be sufficient (I have never seen a report of an issue like this, but it's worth noting that the above suggestion does not guarantee consistency).

## Proposed Solution

Right now `packages.yaml` externals looks like:

```
packages:
  Y:
    externals:
    - spec: Y@1.0
      prefix: /path/to/Y1/
```

If we updated it as follows it would contain sufficient information to construct a DAG:

```
packages:
  Y:
    externals:
    - spec: Y@1.0
      prefix: /path/to/pkg1/
      depends_on: [z1]
  Z:
    externals:
    - spec: Z@1.0
      prefix: /path/to/Z1/
      id: z1
```

in other words, a user can attach an arbitrary (but unique) ID to any external spec entry, and other external specs can claim to depend on it. Specifying ID's and dependencies is not required (in which case Spack behaves as before).

Potential issues:

* ID name collisions (particularly across scopes): Spack should prefer dependencies in the same scope
* Old concretizer runs into more constraint issues: generally externals are chosen as soon as possible, but conditional dependencies can delay this

**Comment by scheibelp** [comment]
@tgamblin

**Comment by alalazo** [comment]
We have been discussing with @tgamblin that configuring Spack to either:
- Prefer packages that are already installed
- Prefer binary packages in a build cache
- Prefer external packages

basically means instructing the ASP based solver to prefer an already concrete spec if possible. The rationale is that in all of the three cases Spack cannot modify something, but can just reuse it as is. So, here comes a wild idea. 

In the discussion above we are trying to modify `packages.yaml` to be able to encode graphs. My thought is that we already have a structure which is very good at this job: **the JSON database**. So I was wondering if:
- Performing external detection only via CLI
- Constructing DB entries from external detection directly (without the need to install them separately afterwards)

could be a possible way forward. Roughly I envision people looking for external tools or libraries on a system only via:
```console
$ spack external find ...
```
The entries that are now in `packages.yaml` (and duplicated in the DB once something is "installed") would appear directly in some DB after `spack external find` finishes. At a high-level I think an implementation of this kind is possible, since:
- The protocol for finding externals permits to determine all the attributes of the node
- The metadata in the package being installed permits to recurse the search for external nodes, using the same protocol, on dependencies

We'll probably need to write code to find externals based on module files (we don't actually have it) if we go this route. What I find appealing of this idea is that:
1. It reduces storing the representation of externals with dependencies to a known problem
2. It automates the detection of externals (so no more a `packages.yaml` where part of the file is written by Spack, part by the user)
3. If both externals and installed packages are stored in the DB, we can reuse the same mechanism to prefer one or the other when solving constraints

**TL;DR** Why don't we try to reduce external detection to the construction of a DB containing external packages and leveraging the current detection protocol (instead of extending `packages.yaml` to represent graphs)?

**Reply by scheibelp** [reply]
> Roughly I envision people looking for external tools or libraries on a system only via `spack external find`

I think `packages.yaml` has advantages:

* Requires minimal knowledge of Spack vs. the potential work of implementing needed methods to integrate with `spack compiler find`
* Under complete control of the user: they get to specify exactly what they want, vs. arranging for a tool to (maybe) find what they want

Even if we can sort out how to get `spack external find` to generate the appropriate database entries, based on the above I think `packages.yaml` will always be useful.

As a side note: updating `spack external find` to generate database entries would also require updating the concretizer. I think that would be fairly straightforward although this assumes that we would be placing concretized instances of the external spec in the DB (right now the entries in packages.yaml are not concretized, which is generally to simplify their integration into the concretization process).

> We'll probably need to write code to find externals based on module files (we don't actually have it)

Module files may not always be available (or correct). I think this would be convenient and helpful as a follow-up.

When module files are not available, properly detecting dependencies is likely more of a challenge. Presumably you could concretize the external spec, then look in the same prefix for instances of dependency libraries etc.

**Reply by scheibelp** [reply]
One thing that came up today: generally externals are "fully concretized" as part of their selection when concretizing a dependent: any unspecified variants or properties of the spec are inferred to be whatever a dependent needs. This is generally to avoid forcing the user to specify everything about the external package. I expect that many users benefit from this flexibility (although this can lead to externals being used when not appropriate).

If externals were to be directly recorded in the database, we'd likely benefit from storing them without automatically assigning any of their unspecified properties.

**Comment by tgamblin** [comment]
@scheibelp: I like this proposal, at least as far as `packages.yaml` syntax goes.  Having an `id` for each external makes sense -- we need a way to reference other externals in `packages.yaml`.

On the question that @alalazo raised: I tend to agree that we need both `packages.yaml` and the spec format in the DB.  One's for humans to specify externals "abstractly" -- i.e., just provide what they know about a package.  One is really intended only for concrete specs.  we don't *really* have a great spec format for abstract specs (or one where it's easy to write just part of the spec) so I think we should stick to `packages.yaml` for user provided stuff right now.

Random thought: I could also see users wanting to use the `id` to refer to specific externals on the CLI (e.g., they could use it like a hash, `foo^/z1` and then have that translated to a specific hash), but that can come later. Right now it is tricky, beyond `buildable: False` and some disambiguation, to refer to one particular external.  This would fix that.

I think @scheibelp should go ahead with this but I want to emphasize the priorities. I think `packages.yaml` format is actually second priority here. The things we need to get done for externals with dependencies are:

1. Make Spack ingest `spack.json` from the Cray PE, and have it create external DB entries from that.  Basically, `spack.json` needs to be a way of specifying available externals (and compilers).  It is more detailed -- closer to a concrete spec -- than what is in `packages.yaml`  It actually has a lot of dependency info.  I think we should do this and the internal model for externals with dependencies *first*.
2. Externals with dependencies in `packages.yaml` -- add `id`, allow users to specify externals that depend on one another, etc.  This is really just an enhancement to `packages.yaml` once we have (1).

The questions I still have about the semantics here are:
- What about dependency types? How are those inferred? Should they be specifiable or always inferred from current `packages.yaml`?  Do we ever need a hint from the user?
- What do we do when dependency information is partial?  On the command line, `foo ^bar` means "foo somehow depends on bar". i.e., both `foo -> baz ->  bar` and `foo -> bar` satisfy the constraint. I think that is probably the abstract interpretation you want for dependencies in `packages.yaml`, as well... I could be persuaded that I am wrong about this, but I bet users likely don't want to write out every intermediate dependency for all packages -- just the ones they know they need. How do we handle that when translating things to concrete specs?  Do we care?

So, given all that, @scheibelp can you take a stab at figuring out the *internal* changes needed for (1)?  And also take a stab at getting the parser and/or `external find` integration working for `spack.json`?

**Reply by scheibelp** [reply]
I want to clarify your comment on https://github.com/spack/spack/discussions/21636
 
> 1. Make Spack ingest spack.json from the Cray PE, and have it create external DB entries from that. Basically, spack.jsonneeds to be a way of specifying available externals (and compilers). It is more detailed -- closer to a concrete spec -- than what is in packages.yaml It actually has a lot of dependency info. I think we should do this and the internal model for externals with dependencies first.
> 2. Externals with dependencies in packages.yaml -- add id, allow users to specify externals that depend on one another, etc. This is really just an enhancement to packages.yaml once we have (1).
 
A few points:
 
* As I understand it, these suggestions are significantly different from what either @alalazo or I proposed - not saying that's a problem, but nevertheless want to clarify since it isn’t stated up front that you want something different
* It sounds like for [1] that you want spack.json specifically to be converted directly to database entries vs. entries in packages.yaml
  * The difference between this and what @alalazo suggested is that you are only proposing to do this for spack.json and not for all externals
* as stated, [2] and [1] can be done completely separately (i.e. if the spack.json data is directly used to create DB entries, then those already record dependencies and there is no need for [2] to support Cray)

Assuming I understood correctly that we want to convert directly to database entries, I have a few follow-up questions:
 
* How are users expected to refer to (and use) these packages once they are added to the database? One method that occurs to me: once they are added to the DB, then spack find would report their hash, and users could refer to those to use the dependencies (i.e. spack spec foo ^/hashofreadpackage).
  * I'm not sure if that would be an intuitive means to use them though (vs. setting buildable: False in packages.yaml)
* One of the points about reading into the DB vs. packages.yaml is concretizing the specs: one advantage of packages.yaml for referring to system-installed packages is that you don't have to specify all variants on the package
  * You could leave variants unspecified but concrete in the external DB entries, but that leaves them less-usable (since dependents requesting certain properties wouldn’t be able to use the external specs in the DB)
 
Overall, you note that spack.json provides enough details to get something closer to a full spec, but the above concerns would generally lead me to using packages.yaml. Nevertheless, the above also outlines how these packages would be used in this paradigm, so if you prefer that I can proceed with it.

**Reply by tgamblin** [reply]
> It sounds like for [1] that you want spack.json specifically to be converted directly to database entries vs. entries in packages.yaml
> * The difference between this and what @alalazo suggested is that you are only proposing to do this for spack.json and not for all externals

This is accurate.  The reason we want to do this is because they're maintaining `spack.json` for us, in a format close to what we already have in the database.  We should exploit that (and if there end up being a lot of packages in the PE, the JSON is faster to read in).

The PE reuse is top priority right now -- we need to get this in by September for the next quarterly, as users are starting to get on machines that have this in the PE.

> How are users expected to refer to (and use) these packages once they are added to the database? One method that occurs to me: once they are added to the DB, then spack find would report their hash, 

I agree with you that this is not intuitive, but I also think reuse by hash is the thing to do first (it's simple, and we have to support it anyway).

I am currently working on reuse of installed packages, which will make the concretizer optimize to reuse what's already on the system.  What I *think* will become intuitive is to let the concretizer fill in the installed packages based on simpler constraints you provide on the CLI or in your environment.  So, for example, you'll just say you prefer the cray compilers, and you'll get the versions of dependencies built with those from the PE via optimization in the solver.

If we need to add other mechanisms (e.g., ids like you are suggesting for `packages.yaml`) then I think we can add that later, but my intent is to line this up with optimizing for reuse with the new concretizer.

> You could leave variants unspecified but concrete in the external DB entries, but that leaves them less-usable (since dependents requesting certain properties wouldn’t be able to use the external specs in the DB)

There is probably some larger discussion on what our policy should really be here. One of the reasons we want externals to have dependencies is so that we can better handle externals like ROCm, where the packages frequently need to refer to their dependencies.  It's common to use an external ROCm/CUDA/etc., but many packages have workarounds in them right now because accusing known dependencies through the spec doesn't work for externals. This will get that working.

Missing details about variants, etc. are similar in that they can prevent packages from being portable.  I think that's  a larger discussion, esp. for `packages.yaml`.  For the Cray-provided `spack.json`, I think we can ask for any needed variants, and they'll be provided.  It avoids the issue of coming up with a syntax in `packages.yaml`, at least for now, and divides this task up into two pieces.

> You could leave variants unspecified but concrete in the external DB entries, but that leaves them less-usable (since dependents requesting certain properties wouldn’t be able to use the external specs in the DB)

Yep please do if you don't see any issues.  @alalazo would be good to get your thoughts too.

**Reply by alalazo** [reply]
I just finished reading through the discussion. I agree with leaving variants out or unspecified in the DB if they are externals, since a missing information there means we cannot grant that a certain specific request on the variant can be honored. 

I also see this lining up well with the discussion on `use_variants`. Right now if we have to detect e.g. `mkl` we might need to put N external specs in the DB, while once `use_variant` will be introduced we'll have a DB that represents better the reality of deployments.

>  One of the reasons we want externals to have dependencies is so that we can better handle externals like ROCm, where the packages frequently need to refer to their dependencies. It's common to use an external ROCm/CUDA/etc., but many packages have workarounds in them right now because accusing known dependencies through the spec doesn't work for externals

Your discussion above on what should be extended in `packages.yaml` and what should be directly stored in the DB makes me think of a slight modification of my initial proposal. What if we:
1. Extend the detection semantic for packages to go to dependencies
2. Register whatever is found by `spack external find` directly in the DB 

Point 1 should make it possible to deal with things like ROCm, provided people coded the support for that in the packages. Point 2 separates machine generated externals from the ones written manually by the user. The extension proposed by Peter in `packages.yaml` would be instead a way to work around packages that don't have detection coded already or that have a detection mechanism that needs improvement. Would something like this make sense in the long run, or do you see issues with that?

**Reply by alalazo** [reply]
> What about dependency types? How are those inferred? Should they be specifiable or always inferred from current packages.yaml? Do we ever need a hint from the user?

I don't know how it could be represented in the DB or in a spec, but when dealing with externals we might have situations in which something is detected and can be used only with certain dependency types. For instance, in many distros what is considered a single package in Spack is divided in many system packages e.g.
- `devel` containing the headers
- `libs` containing the libraries
- `app` containing executables

Having only part of them installed might mean that only certain kind of dependencies can be satisfied by the external (e.g. if we have only the `app` we can satisfy only `run` type dependencies). For most packages we can probably have good heuristics to detect automatically if headers, libraries or apps are installed.

To what concerns instead the dependencies of the external _being detected_, I think the only one we are insterested in are direct link and run dependencies, like in the case of a buildcache.

**Reply by tgamblin** [reply]
> 1. Extend the detection semantic for packages to go to dependencies
> 2. Register whatever is found by spack external find directly in the DB
> 
> Point 1 should make it possible to deal with things like ROCm, provided people coded the support for that in the packages. Point 2 separates machine generated externals from the ones written manually by the user. 

I think if we can make this seamless with `spack external find`, then that would be good.  What is not clear to me is how we work in a new mechanism there -- specifically what we're targeting here is the `spack.json`, which looks like a DB.  It seems like packages should have a say in interpreting that, via the `spack external find` mechanism, but I don't know how it should be implemented.  Right now we find a bunch of executables and hand them to `determine_spec_details`.  Do we use `spack.json` to find executables, but also pass the JSON record for, e.g., `hdf5` to `hdf5`'s `determine_spec_details` method?  that might make sense.  You could mostly use the JSON and fill in blanks by interrogating the executable.

> The extension proposed by Peter in packages.yaml would be instead a way to work around packages that don't have detection coded already or that have a detection mechanism that needs improvement.

Yep -- though I suspect that anything in `packages.yaml` may need to come after we get the initial feature implemented.

**Reply by alalazo** [reply]
> It seems like packages should have a say in interpreting that, via the spack external find mechanism, but I don't know how it should be implemented.

Just thinking out loud here. Right now we have a single mechanism based on searching `${PATH}` + a method that gets called on packages to determine versions, variants etc. I think the first step might be to add a new mechanism based on `spack.json`:
```console
$ spack external find --method=cray-json ...
```
That method should trigger a different workflow where the information from `spack.json` is first passed to a method similar in principle to `determine_spec_details` but specific to this way of searching software on Cray. The specs returned by this method are then installed directly into the store, and marked in a way that permits to differentiate them. This is so that we can add an additional subcommand:
```console
spack external remove
``` 
to permit reverting the previous operation.
